UI kit input

JorgeAMendoza opened the pull request:
Create `Input` element from the UI kit, apply various stylings and states and implement the component into the `filter` options.

JorgeAMendoza committed:
remove comment

JorgeAMendoza committed:
input kit configuration

create component and various storybook components to be displayed when
running storybook. Basic states and options rendered with docs.

JorgeAMendoza committed:
pass in correct args

ensure that two storybook components properly receive the args.

JorgeAMendoza committed:
input styles

set styles for the no icon, no error components, set style classes in
the scss file.

JorgeAMendoza committed:
input icon

adjust styles for inputs that will contain an icon. Also set the focus
state to have purple border.

JorgeAMendoza committed:
label input

style label on the input component and ensure that styles render on
storybook.

JorgeAMendoza committed:
hint text

style the hint text in the `Input` component.

JorgeAMendoza committed:
error state

style the input for the error state, styling the error text, icon, and
label border.

JorgeAMendoza committed:
Implement `Input` component

Implement the `Input` UI Kit into the filter options.

romanchukenator commented on features/ui/input/Input.tsx:
```diff
@@ -0,0 +1,56 @@
+import style from "./input.module.scss";
+
+interface InputProps {
+  error?: boolean;
+  errorText?: string;
```
You could likely just use the `errorText` to figure if there's an error or not.

JorgeAMendoza replied:
According to the Figma design, an input can have an error and not error text, wouldn't just using `errorText` make it difficult to render the "no error text but error" input style?

romanchukenator replied:
Oh, interesting. Right you are. Sorry about that.

I would push back about not showing an error message but the designs say otherwise! 🤷

romanchukenator commented on features/ui/input/Input.tsx:
```diff
@@ -0,0 +1,56 @@
+        {icon ? <img src={icon} alt="" className={style.inputIcon} /> : null}
+        <input {...inputProps} type="text" />
+
+        {error ? (
+          // eslint-disable-next-line @next/next/no-img-element
+          <img
+            src="/icons/alert-circle.svg"
+            alt=""
+            className={style.errorIcon}
+          />
+        ) : null}
+      </span>
+
+      {hint && <span className={style.hintText}>{hint}</span>}
+      {error && <span className={style.errorText}>{errorText}</span>}
```
These still work with `{errorText && <span className={style.errorText}>{errorText}</span>}`, etc.

JorgeAMendoza replied:
It would probably be best to check against `errorText` instead of `error` when determining to render the error text. This will be changed.

romanchukenator reviewed (changes requested):
Just one thing about using truthy/falsey values.

romanchukenator commented on features/ui/input/Input.tsx:
```diff
@@ -0,0 +1,56 @@
+}: InputAttributes & InputProps) => {
+  return (
+    <label
+      aria-label={!label ? ariaText : ""}
```
```suggestion
      aria-label={label ? "" : ariaText}
```

It's kinda 🌵 a nitpick but I personally try to avoid inverting checked values because of the additional cognitive overhead of having to read lines of code while keeping track of things like "if this is true then it's actually false"

JorgeAMendoza replied:
I agree with this, sometimes I invert the checks mistakenly! Will be committed.

romanchukenator commented on features/ui/input/Input.tsx:
```diff
@@ -0,0 +1,56 @@
+      data-disabled={inputProps.disabled}
+      data-error={error}
+    >
+      {label ? <span className={style.labelText}>{label}</span> : null}
```
```suggestion
      {label && <span className={style.labelText}>{label}</span>}
```

romanchukenator replied:
...and the same comment for any place you use a ternary check to possibly resolve to `null`

JorgeAMendoza replied:
I prefer to use the ternary operator in React because I want to be deliberate about what to render in the component. By stating null, react knows not to render anything, but with `&&`, react my potentially render the component. In this case its not a big deal, but if the element was going to be rendered based on a API call, `&&` would throw an error on the initial render because it does not have the data yet! I just realized I mixed how I rendered text this way and need to make it consistent!

I hope I explained it right, here is a article detailing the differences: https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx

JorgeAMendoza committed:
comment updates

update storybook names to make the story clearer, make changes based on
feedback of comment.

mentor-tara reviewed (approved):
Go go go :smile:[](prolog-21)
